File: rust-examples/src/patternmatching.rs
enum QuadraticSolution {
    NoSolution,
    OneSolition(f64),
    TwoSolutions(f64, f64),
}

fn solve_quadratic_equation(a: f64, b: f64, c: f64) -> QuadraticSolution {
    let discriminant = b * b - 4.0 * a * c;
    if discriminant <  0.0 { QuadraticSolution::NoSolution }
    else if discriminant == 0.0 { QuadraticSolution::OneSolition(-b / (2.0 * a))} 
    else { QuadraticSolution::TwoSolutions((-b + discriminant.sqrt()) / (2.0 * a), (-b - discriminant.sqrt()) / (2.0 * a))}
}

fn main() {
    let res = solve_quadratic_equation(1.0, 3.0, 1.0);
    match res {
        QuadraticSolution::NoSolution => println!("No solution"),
        QuadraticSolution::OneSolition(x) => println!("One solution: {}", x),
        QuadraticSolution::TwoSolutions(x,y) => println!("Two solutions: {}, {}", x,y)
    }

    let res2 = if let QuadraticSolution::NoSolution = solve_quadratic_equation(1.0, 1.0, 1.0) {
        5.0
    }
    else {
        8.0
    };

    println!("{}",res2);
}

File: cpp-examples/patternmatching.cpp
#include <variant>
#include <iostream>
#include <utility>
#include <cmath>

template <class... Ts>
struct overloaded : Ts...
{
    using Ts::operator()...;
};
template <class... Ts>
overloaded(Ts...) -> overloaded<Ts...>;

using quadratic_equation_solution = std::variant<std::monostate, double, std::pair<double, double>>;

quadratic_equation_solution solve_quadratic_equation(double a, double b, double c)
{
    const auto discriminant = b * b - 4.0 * a * c;

    if (discriminant < 0.0)
        return std::monostate{};
    else if (discriminant == 0.0)
        return -b / (2.0 * a);
    else
    {
        return std::pair{(-b + std::sqrt(discriminant)) / (2.0 * a),
                         (-b - std::sqrt(discriminant)) / (2.0 * a)};
    }
}

int main()
{
    const auto res = solve_quadratic_equation(1.0, 3.0, 1.0);

    std::visit(overloaded{
                   [](std::monostate arg) { std::cout << "No solution!" << '\n'; },
                   [](double arg) { std::cout << "One solution: " << arg << 'n'; },
                   [](std::pair<double, double> arg) { std::cout << "Two solutions: "
                                                                 << arg.first << " " << arg.second << '\n'; },
               },
               res);
}